File: app-phone/src/androidTest/java/com/code2hack/eyebrowse/phone/SingleShotSwipe.java
package com.code2hack.eyebrowse.phone;

import android.os.SystemClock;
import android.view.MotionEvent;
import android.view.ViewConfiguration;

import androidx.test.espresso.InjectEventSecurityException;
import androidx.test.espresso.UiController;
import androidx.test.espresso.action.MotionEvents;
import androidx.test.espresso.action.Press;

import java.util.ArrayList;
import java.util.List;

/**
 * Test-only one-attempt Espresso-compatible pointer construction on an already-captured UiController.
 * GeneralSwipeAction/MotionEvents retry wrappers are not used. A false controller result is a hard
 * single-attempt failure; checked injection failures keep their original object identity.
 */
final class SingleShotSwipe {
    private static final int SWIPE_EVENT_COUNT = 10;
    private static final int FAST_DURATION_MS = 150;

    private SingleShotSwipe() {}

    static MotionEvent obtainTapDown(InputSafety.Path point) {
        return MotionEvents.obtainDownEvent(
                new float[] {point.startX, point.startY}, Press.FINGER.describePrecision());
    }

    /**
     * Sends one Espresso-shaped DOWN and gives the main loop the same bounded tap-detection dwell
     * used by MotionEvents.sendDown, but never retries a false/uncertain DOWN.
     */
    static void injectTapDown(UiController controller, MotionEvent down) {
        requireController(controller);
        try {
            if (!controller.injectMotionEvent(down)) {
                throw new IllegalStateException("Espresso tap DOWN injection returned false");
            }
            long isTapAt = down.getDownTime() + (ViewConfiguration.getTapTimeout() / 2L);
            while (true) {
                long delay = isTapAt - SystemClock.uptimeMillis();
                if (delay <= 10) {
                    break;
                }
                controller.loopMainThreadForAtLeast(Math.max(1L, delay / 4L));
            }
        } catch (InjectEventSecurityException original) {
            SingleShotSwipe.<RuntimeException>sneakyThrow(original);
        }
    }

    static void injectTapUp(UiController controller, MotionEvent down, InputSafety.Path point) {
        requireController(controller);
        MotionEvent up = MotionEvents.obtainUpEvent(
                down, new float[] {point.endX, point.endY});
        try {
            try {
                if (!controller.injectMotionEvent(up)) {
                    throw new IllegalStateException("Espresso tap UP injection returned false");
                }
            } catch (InjectEventSecurityException original) {
                SingleShotSwipe.<RuntimeException>sneakyThrow(original);
            }
        } finally {
            up.recycle();
        }
    }

    /**
     * Exact pinned FAST shape: DOWN, ten linear MOVE points, UP over 150 ms. Unlike Swipe.FAST,
     * this helper observes injectMotionEventSequence's boolean result instead of silently returning
     * SUCCESS when InputManager reports false.
     */
    static void send(UiController controller, InputSafety.Path path) {
        requireController(controller);
        float[] start = new float[] {path.startX, path.startY};
        float[] end = new float[] {path.endX, path.endY};
        float[] precision = Press.FINGER.describePrecision();
        List<MotionEvent> events = new ArrayList<>();
        MotionEvent down = MotionEvents.obtainDownEvent(start, precision);
        events.add(down);
        try {
            long intervalMs = FAST_DURATION_MS / SWIPE_EVENT_COUNT;
            long eventTime = down.getDownTime();
            for (int i = 1; i <= SWIPE_EVENT_COUNT; i++) {
                eventTime += intervalMs;
                float fraction = i / (SWIPE_EVENT_COUNT + 2.0f);
                float[] point = new float[] {
                        start[0] + (end[0] - start[0]) * fraction,
                        start[1] + (end[1] - start[1]) * fraction
                };
                events.add(MotionEvents.obtainMovement(down, eventTime, point));
            }
            eventTime += intervalMs;
            events.add(MotionEvents.obtainUpEvent(down, eventTime, end));
            try {
                if (!controller.injectMotionEventSequence(events)) {
                    throw new IllegalStateException("Espresso swipe injection returned false");
                }
            } catch (InjectEventSecurityException original) {
                SingleShotSwipe.<RuntimeException>sneakyThrow(original);
            }
        } finally {
            for (MotionEvent event : events) {
                event.recycle();
            }
        }
    }

    private static void requireController(UiController controller) {
        if (controller == null) {
            throw new IllegalStateException("captured Espresso UiController unavailable");
        }
    }

    /** Preserve the exact checked Espresso injection throwable through Runnable-based dispatch. */
    @SuppressWarnings("unchecked")
    private static <T extends Throwable> void sneakyThrow(Throwable original) throws T {
        throw (T) original;
    }
}

File: app-phone/src/androidTest/java/com/code2hack/eyebrowse/phone/SingleShotSwipe.kt
package com.code2hack.eyebrowse.phone

import android.os.SystemClock
import android.view.MotionEvent
import android.view.ViewConfiguration
import androidx.test.espresso.action.MotionEvents
import androidx.test.espresso.action.Press
import androidx.test.platform.app.InstrumentationRegistry

/**
 * Test-only single-submission pointer injection through Instrumentation.
 *
 * Event construction retains the pinned Espresso tap/FAST gesture shape, but the actual system
 * submission does not enter Espresso's retrying InputManagerEventInjectionStrategy. Every generated
 * event is submitted at most once and is never resubmitted. Android 12 sendPointerSync is void:
 * thrown failures are visible to this harness, while a false system injection result (and a
 * RemoteException swallowed inside Instrumentation) is not observable at submission time.
 */
internal object SingleShotSwipe {
    private const val SWIPE_EVENT_COUNT = 10
    private const val FAST_DURATION_MS = 150

    fun obtainTapDown(point: InputSafety.Path): MotionEvent =
        MotionEvents.obtainDownEvent(
            floatArrayOf(point.startX, point.startY),
            Press.FINGER.describePrecision(),
        )

    /**
     * One DOWN submission. The following sleep is only the retained tap-detection gesture cadence,
     * not readiness observation, retry, or rejection recovery.
     */
    fun injectTapDown(down: MotionEvent) {
        instrumentation().sendPointerSync(down)
        sleepUntil(down.downTime + ViewConfiguration.getTapTimeout() / 2L)
    }

    fun injectTapUp(
        down: MotionEvent,
        point: InputSafety.Path,
    ) {
        val eventTime = maxOf(
            SystemClock.uptimeMillis(),
            down.downTime + ViewConfiguration.getTapTimeout() / 2L,
        )
        val up = MotionEvents.obtainUpEvent(
            down,
            eventTime,
            floatArrayOf(point.endX, point.endY),
        )
        try {
            instrumentation().sendPointerSync(up)
        } finally {
            up.recycle()
        }
    }

    /**
     * Exact pinned FAST shape: DOWN, ten linear MOVE points, UP over 150 ms. Each event is
     * submitted once in order through Instrumentation.sendPointerSync and is never resubmitted.
     * A thrown failure aborts immediately; silent platform-side failure is not observable here and
     * is therefore detected only by the caller's downstream scroll/effect assertions.
     */
    fun send(path: InputSafety.Path) {
        val start = floatArrayOf(path.startX, path.startY)
        val end = floatArrayOf(path.endX, path.endY)
        val precision = Press.FINGER.describePrecision()
        val events = mutableListOf<MotionEvent>()
        val down = MotionEvents.obtainDownEvent(start, precision)
        events += down
        try {
            val intervalMs = FAST_DURATION_MS / SWIPE_EVENT_COUNT
            var eventTime = down.downTime
            for (i in 1..SWIPE_EVENT_COUNT) {
                eventTime += intervalMs.toLong()
                val fraction = i / (SWIPE_EVENT_COUNT + 2.0f)
                val point = floatArrayOf(
                    start[0] + (end[0] - start[0]) * fraction,
                    start[1] + (end[1] - start[1]) * fraction,
                )
                events += MotionEvents.obtainMovement(down, eventTime, point)
            }
            eventTime += intervalMs.toLong()
            events += MotionEvents.obtainUpEvent(down, eventTime, end)

            for (event in events) {
                sleepUntil(event.eventTime)
                instrumentation().sendPointerSync(event)
            }
        } finally {
            events.forEach(MotionEvent::recycle)
        }
    }

    private fun instrumentation() = InstrumentationRegistry.getInstrumentation()

    private fun sleepUntil(eventTime: Long) {
        val remaining = eventTime - SystemClock.uptimeMillis()
        if (remaining > 0) {
            SystemClock.sleep(remaining)
        }
    }
}
